src/domain/task-system/src/aggregates/task.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Utc};
+use enum_variants::*;
+use event_sourcing::*;
+
+use crate::entities::*;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+pub struct Task {
+    state: TaskState,
+    pending_events: Vec<TaskEvent>,
+    last_synced_event: Option<EventID>,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl Task {
+    pub fn new(task_id: TaskID, logical_plan: LogicalPlan) -> Self {
+        let genesis = TaskCreated {
+            event_time: Utc::now(),
+            task_id,
+            logical_plan,
+        };
+
+        Self {
+            state: TaskState {
+                task_id,
+                status: TaskStatus::Queued,
+                cancellation_requested: false,
+                logical_plan: genesis.logical_plan.clone(),
+                created_at: genesis.event_time.clone(),
+                ran_at: None,
+                cancellation_requested_at: None,
+                finished_at: None,
+            },
+            pending_events: vec![genesis.into()],
+            last_synced_event: None,
+        }
+    }
+
+    /// Transition task to a `Running` state
+    pub fn run(&mut self) -> Result<(), IllegalSequenceError<Self>> {
+        self.apply(TaskRunning {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+        })
+    }
+
+    /// Task is queued or running and cancellation was not already requested
+    pub fn can_cancel(&mut self) -> bool {
+        match self.status {
+            TaskStatus::Queued if !self.cancellation_requested => true,
+            TaskStatus::Running if !self.cancellation_requested => true,
+            _ => false,
+        }
+    }
+
+    /// Set cancellation flag (if not already set)
+    pub fn cancel(&mut self) -> Result<(), IllegalSequenceError<Self>> {
+        if self.cancellation_requested {
+            return Ok(());
+        }
```

Comment:
Why is this check needed? I guess if we have multiple cancellations in the queue of events, well, the 2nd one should just silently ignore

Reply:
While it's legal to cancel already cancelled task I though it's better to save some work of writing a redundant event.